# Unstructured PDF Extractuion Pipeline Setup and Testing
This file serves as a playground/workspace to demo Unstructured functionality before it is put into production. This is where we will test the extraction, chunking, metadata extraction, metadata tagging, and vector embedding of PDFs for CASSIE

#### Changelog
**May 18th, 2024**  
**7:00 AM**  
* File created  
* Metadata extraction testing  
**7:30 AM**
* PDF Element Extraction

### Metadata Extraction from PDF Files

#### Imports

In [116]:
import pikepdf
import sys

#### Initial Extraction of Metadata

In [117]:
pdf_file = 'C:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs/Security Docs/5000-00-CTSC-48PA-0004_01.pdf'

pdf = pikepdf.Pdf.open(pdf_file)
docInfo = pdf.docinfo

for key, value in docInfo.items():
    print(f'{key}: {value}')

/CreationDate: D:20181004144423-04'00'
/Creator: Microsoft® Word 2013
/ModDate: D:20181004155108-04'00'
/Producer: Microsoft® Word 2013


#### Transform Date Function

In [118]:
import pikepdf
import datetime
import re
from dateutil.tz import tzutc, tzoffset
import sys

In [119]:
pdf_date_pattern = re.compile(''.join([
    r"(D:)?",
    r"(?P<year>/d/d/d/d)",
    r"(?P<month>/d/d)",
    r"(?P<day>/d/d)",
    r"(?P<hour>/d/d)",
    r"(?P<minute>/d/d)",
    r"(?P<second>/d/d)",
    r"(?P<tz_offset>[+-zZ])?",
    r"(?P<tz_hour>/d/d)?",
    r"'?(?P<tz_minute>/d/d)?'?"]))

In [120]:
def transform_date(date_str):
    """
    Convert a pdf date such as "D:20120321183444+07'00'" into a usable datetime
    http://www.verypdf.com/pdfinfoeditor/pdf-date-format.htm
    (D:YYYYMMDDHHmmSSOHH'mm')
    :param date_str: pdf date string
    :return: datetime object
    """
    global pdf_date_pattern
    match = re.match(pdf_date_pattern, date_str)
    if match:
        date_info = match.groupdict()

        for k, v in date_info.items():  # transform values
            if v is None:
                pass
            elif k == 'tz_offset':
                date_info[k] = v.lower()  # so we can treat Z as z
            else:
                date_info[k] = int(v)

        if date_info['tz_offset'] in ('z', None):  # UTC
            date_info['tzinfo'] = tzutc()
        else:
            multiplier = 1 if date_info['tz_offset'] == '+' else -1
            date_info['tzinfo'] = tzoffset(None, multiplier*(3600 * date_info['tz_hour'] + 60 * date_info['tz_minute']))

        for k in ('tz_offset', 'tz_hour', 'tz_minute'):  # no longer needed
            del date_info[k]

        return datetime.datetime(**date_info)

#### Extract Dates from PDF

In [121]:
pdf = pikepdf.Pdf.open(pdf_file)
docinfo = pdf.docinfo
for key, value in docinfo.items():
    if str(value).startswith("D:"):
        # pdf datetime format, convert to python datetime
        value = transform_date(str(pdf.docinfo["/CreationDate"]))
    print(key, ":", value)

/CreationDate : None
/Creator : Microsoft® Word 2013
/ModDate : None
/Producer : Microsoft® Word 2013


So it looks like the created and modified dates can be extracted from PDFs, even though I've downloaded them much later. I will try this for all the PDF files in the CSDocs folder to ensure it works properly

In [122]:
import os

In [123]:
CSDocs_path = 'C:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs/'

doc_info = []


for file in os.listdir(CSDocs_path):
        try:
            doc = {}
            if file.endswith('.pdf'):
                doc = {'filename': file}
                pdf = pikepdf.Pdf.open(CSDocs_path + file)
                docInfo = pdf.docinfo
                for key, value in docInfo.items():
                    print(f'{key}: {value}')
                    if str(value).startswith("D:"):
                        # pdf datetime format, convert to python datetime
                        value = transform_date(str(pdf.docinfo["/CreationDate"]))
                    if key == '/CreationDate':
                        key = 'Creation Date'
                    if key == '/Creator':
                        key = 'Creator'
                    if key == '/ModDate':
                        key = 'Modification Date'
                    if key == '/Producer':
                        key = 'Producer'
                    if key == '/Title':
                        key = 'Title'
                    if key == '/Subject':
                        key = 'Subject'
                    if key == '/Author':
                        key = 'Author'
                    doc[key] = value
                print('\n')
            if doc['filename']:
                doc_info.append(doc)
        except Exception as e:
            print(e)

/Author: Yousef Kimiagar
/CreationDate: D:20200210124719-05'00'
/Creator: Microsoft® Word for Office 365
/ModDate: D:20200210124719-05'00'
/Producer: Microsoft® Word for Office 365
/Title: SCHEDULE 15-2


/Author: David Robson/Duncan Robb
/CreationDate: D:20170927074532-04'00'
/Creator: Microsoft® Word 2013
/ModDate: D:20170927083115-04'00'
/Producer: Microsoft® Word 2013
/Subject: Eglinton Crosstown LRT Project
/Title: ECLRT Cyber Security Management Plan


'filename'
/Author: 
/CreationDate: D:20220506210237-04'00'
/Creator: Aspose Ltd.
/ModDate: D:20220506210237-04'00'
/Producer: Aspose.Pdf for .NET 11.4.0
/Subject: 
/Title: 


/CreationDate: D:20210831101215-04'00'
/Creator: Microsoft® Word for Microsoft 365
/ModDate: D:20220224143709-05'00'
/Producer: Microsoft® Word for Microsoft 365


/CreationDate: D:20210714131927-04'00'
/Creator: Microsoft® Word for Microsoft 365
/ModDate: D:20220224145956-05'00'
/Producer: Microsoft® Word for Microsoft 365


/CreationDate: D:20210715181444-0

In [124]:
for i in doc_info:
    for key, value in i.items():
        print(f'{key}: {value}')
    print('\n')



filename: 15-2 Conformed.pdf
Author: Yousef Kimiagar
Creation Date: None
Creator: Microsoft® Word for Office 365
Modification Date: None
Producer: Microsoft® Word for Office 365
Title: SCHEDULE 15-2


filename: 5000-00-CTSC-48PA-0002 - Cyber Security Management Plan-5000-00-CTSC-48PA-0002_PA.pdf
Author: David Robson/Duncan Robb
Creation Date: None
Creator: Microsoft® Word 2013
Modification Date: None
Producer: Microsoft® Word 2013
Subject: Eglinton Crosstown LRT Project
Title: ECLRT Cyber Security Management Plan


filename: 5000-00-WGD-48PA- 2001 - Internal Handover Plan – Cyber Systems-2700-67ASA1-02-CTSC-0004_01.pdf
Author: 
Creation Date: None
Creator: Aspose Ltd.
Modification Date: None
Producer: Aspose.Pdf for .NET 11.4.0
Subject: 
Title: 


filename: 5000-00-WGD-48PA-1001 - Baseline Controls Document - Management Controls Policy.pdf
Creation Date: None
Creator: Microsoft® Word for Microsoft 365
Modification Date: None
Producer: Microsoft® Word for Microsoft 365


filename: 5000-

In [125]:
for i in doc_info:
    print(i)

{'filename': '15-2 Conformed.pdf', 'Author': pikepdf.String("Yousef Kimiagar"), 'Creation Date': None, 'Creator': pikepdf.String("Microsoft® Word for Office 365"), 'Modification Date': None, 'Producer': pikepdf.String("Microsoft® Word for Office 365"), 'Title': pikepdf.String("SCHEDULE 15-2")}
{'filename': '5000-00-CTSC-48PA-0002 - Cyber Security Management Plan-5000-00-CTSC-48PA-0002_PA.pdf', 'Author': pikepdf.String("David Robson/Duncan Robb"), 'Creation Date': None, 'Creator': pikepdf.String("Microsoft® Word 2013"), 'Modification Date': None, 'Producer': pikepdf.String("Microsoft® Word 2013"), 'Subject': pikepdf.String("Eglinton Crosstown LRT Project"), 'Title': pikepdf.String("ECLRT Cyber Security Management Plan")}
{'filename': '5000-00-WGD-48PA- 2001 - Internal Handover Plan – Cyber Systems-2700-67ASA1-02-CTSC-0004_01.pdf', 'Author': pikepdf.String(""), 'Creation Date': None, 'Creator': pikepdf.String("Aspose Ltd."), 'Modification Date': None, 'Producer': pikepdf.String("Aspose.P

So it looks like there is a good amount of metadata already stored in the metadata of the PDF. We will aim to enhance this metadata using PDF extraction and element recognition. We'll start by reading a PDF with unstructured and seeing what elements exist on the first page/author block.

It is notable that the title metadata isn't always descriptive. I'm not sure how we can get a document description using an LLM, but I think either RAPTOR (Continuously shrinking documents (Subsection Summaries => Section Summaries => Doc Summaries) could work to shorten the tokens needed, or passing to Gemini (The 1M, soon 2M, token context window should be enough to summarize most documents) could be a good solution.)

We will start with Pulling Titles and Authors which starts with Unstructured extraction

### Unstuctured Extraction

In [126]:
pdf_file = "c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs/5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf"

from unstructured.partition.pdf import partition_pdf

elements = partition_pdf(
    pdf_file,
    languages=['en'],
    hi_res_model_name='yolox',
    infer_table_structure=True,
    strategy='hi_res'
    )

In [127]:
print(len(elements))
metadata = elements[0].metadata.to_dict()
el = elements[1].to_dict()

for element in elements:
    element_dict = element.to_dict()
    if element_dict['type'] == 'Title':
        print(element_dict['text'])
        print('\n')
        first_title_element = element_dict
        break

for key, value in metadata.items():
    print(f'{key}: {value}')

print('\n')
for key, value in first_title_element.items():
    print(f'{key}: {value}')

504
Cyber Security – Change Control and Configuration Management Process


coordinates: {'points': ((192.29166666666666, 310.55555555555543), (192.29166666666666, 532.3611111111111), (749.2361111111112, 532.3611111111111), (749.2361111111112, 310.55555555555543)), 'system': 'PixelSpace', 'layout_width': 1700, 'layout_height': 2200}
last_modified: 2023-03-28T16:34:32
filetype: application/pdf
languages: ['en']
page_number: 1
file_directory: c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs
filename: 5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf


type: Title
element_id: 8df1b47d06638e6081109dc18f6b42d0
text: Cyber Security – Change Control and Configuration Management Process
metadata: {'detection_class_prob': 0.48540282249450684, 'coordinates': {'points': ((322.7745666503906, 599.533203125), (322.7745666503906, 751.4722222222222), (1370.8916015625, 751.4722222222222), (1370.8916015625, 599.533203125)), 'system': 'PixelSpace', 'layout_width': 1700, 'layo

The file has been partitioned using the yolox model. 
filename: 5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf  
elements: 504  
first title element text: Cyber Security – Change Control and Configuration Management Process  
metadata:  
    **coordinates**:   
        {'points': ((192.29166666666666, 310.55555555555543), (192.29166666666666, 532.3611111111111), (749.2361111111112, 532.3611111111111), (749.2361111111112, 310.55555555555543)), 'system': 'PixelSpace', 'layout_width': 1700, 'layout_height': 2200}
    **last_modified**:   
        2023-03-28T16:34:32
    **filetype**:   
        application/pdf
    **languages**:   
        ['en']
    **page_number**:   
        1
    **file_directory**:  
        c:/Users/Dan/OneDrive/Documents/Crosslinx/CSDocs
    **filename**:   
        5000-00-WGD-48PA-1020 - Change Control and Configuration Management Process.pdf

#### Examining Headers and Footers

In [128]:
for i in elements:
    print(i.to_dict()['type'])
    if i.to_dict()['type'] == 'Table':
        print(i.metadata.text_as_html)
    elif i.to_dict()['type'] == 'NarrativeText':
        print(i.to_dict()['text'])
    elif i.to_dict()['type'] == 'Title':
        print(i.to_dict()['text'])
    elif i.to_dict()['type'] == 'ListItem':
        print(i.to_dict()['text'])
    elif i.to_dict()['type'] == 'UncategorizedText':
        print(i.to_dict()['text'])
    else:
        print(i.to_dict()['type'])

Image
Image
Image
Image
Title
Cyber Security – Change Control and Configuration Management Process
Image
Image
Image
Image
Image
Image
NarrativeText
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. Please access the electronic version for the most current information.
Image
Image
UncategorizedText
5000-00-WGD-48PA-1020
UncategorizedText
03
Title
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Title
Doc Owner
NarrativeText
WGD Consulting
Title
DOCUMENT REVISIONS INDEX
Table
<table><thead><th>Revision</th><th>Revision Date</th><th>Author(s)</th><th>Reviewer(s)</th><th>Description of Changes</th></thead><tr><td>PA</td><td>April 8, 2020</td><td>Chris Goulopou

In [129]:
headers_list = [el for el in elements if el.to_dict()['type'] == 'Header']
footers_list = [el for el in elements if el.to_dict()['type'] == 'Footer']

In [130]:
print(f"Number of Headers: {len(headers_list)}")
print(f"Number of Footers: {len(footers_list)}")

Number of Headers: 6
Number of Footers: 4


In [131]:
for i in headers_list:
    print(i.to_dict()['text'])
for i in footers_list:
    print(i.to_dict()['text'])

Doc Owner
Doc Owner
WGD Consulting
Doc Owner
WGD Consulting
Doc Owner
Page | 22 of 25
Document Classification: CTSC/ECLRT Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. Please access the electronic version for the most current information.
Page | 23 of 25
Page | 24 of 25


In [132]:
tables_list = [el for el in elements if el.to_dict()['type'] == 'Table']

for i in tables_list:
    print(i.metadata.text_as_html)

<table><thead><th>Revision</th><th>Revision Date</th><th>Author(s)</th><th>Reviewer(s)</th><th>Description of Changes</th></thead><tr><td>PA</td><td>April 8, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Initial draft for review</td></tr><tr><td>00</td><td>April 24, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Updated to reflect comments from review</td></tr><tr><td>01</td><td>July 9, 2020</td><td>Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td></tr><tr><td>02</td><td>August 27, 2020</td><td>Mark Salsberg</td><td>Chris Goulopoulos</td><td>Minor corrections and edits</td></tr><tr><td>03</td><td>June 29, 2021</td><td>. Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Adjusted wording for responsibilities based on CTS contractual agreements</td></tr></table>
<table><tr><td>PULDOSEC</td><td>.... ee</td><td>eeeeecceecceeceeeces

In [133]:
def count_elements(elements):
    element_count = {}
    for element in elements:
        element_type = type(element).__name__
        if element_type not in element_count:
            element_count[element_type] = 0
        element_count[element_type] += 1
    return element_count

In [134]:
element_count = count_elements(elements)
for i in element_count:
    print(f'{i}: {element_count[i]}')

Image: 37
Title: 104
NarrativeText: 139
Text: 59
Table: 9
ListItem: 146
Header: 6
Footer: 4


In [135]:
titles = [el for el in elements if el.to_dict()['type'] == 'Title']
for i in titles:
    print(i.to_dict()['text'])

Cyber Security – Change Control and Configuration Management Process
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
DOCUMENT REVISIONS INDEX
REVISION AND CONTROL
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
Contents
Appendix A
Appendix B
Appendix C
Appendix D
Appendix E
Appendix F
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
1 Introduction
2 Purpose
3 Objectives
4 Scope and Applicability
Cyber Security – Change Control and Configuration Management Process 2021-03-01
4.1 Applicability
4.2 Exclusions
4.3 Associated Documents
Cyber Security – Change Control and Configuration Management Process 2021-03-01
Doc Owner
5 Roles and Responsibilities
1. Cyber Security Senior Manager
2. Delegates (Functional Delegates)
3. Technical and Compliance Oversight
Change Review Group
4.

In [136]:
from io import StringIO 
from lxml import etree

table_html = tables_list[0].metadata.text_as_html

parser = etree.XMLParser(remove_blank_text=True)
file_obj = StringIO(table_html)
tree = etree.parse(file_obj, parser)
print(etree.tostring(tree, pretty_print=True).decode())

<table>
  <thead>
    <th>Revision</th>
    <th>Revision Date</th>
    <th>Author(s)</th>
    <th>Reviewer(s)</th>
    <th>Description of Changes</th>
  </thead>
  <tr>
    <td>PA</td>
    <td>April 8, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Initial draft for review</td>
  </tr>
  <tr>
    <td>00</td>
    <td>April 24, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Updated to reflect comments from review</td>
  </tr>
  <tr>
    <td>01</td>
    <td>July 9, 2020</td>
    <td>Chris Goulopoulos_</td>
    <td>| Mark Salsberg</td>
    <td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td>
  </tr>
  <tr>
    <td>02</td>
    <td>August 27, 2020</td>
    <td>Mark Salsberg</td>
    <td>Chris Goulopoulos</td>
    <td>Minor corrections and edits</td>
  </tr>
  <tr>
    <td>03</td>
    <td>June 29, 2021</td>
    <td>. Chris Goulopoulos_</td>
    <td>| Mark

In [137]:
from IPython.core.display import HTML
HTML(table_html)

In [138]:
from langchain_openai import ChatOpenAI
from langchain_core.documents import Document
from langchain.chains.summarize import load_summarize_chain

In [139]:
openai_api_key = os.environ.get('OPENAI_API_KEY')

In [140]:
os.environ['OPENAI_API_KEY'] = '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'
llm = ChatOpenAI(temperature=0,openai_api_key=openai_api_key, model_name="gpt-4o")
chain = load_summarize_chain(llm, chain_type="stuff")
chain.invoke([Document(page_content=table_html)])

{'input_documents': [Document(page_content='<table><thead><th>Revision</th><th>Revision Date</th><th>Author(s)</th><th>Reviewer(s)</th><th>Description of Changes</th></thead><tr><td>PA</td><td>April 8, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Initial draft for review</td></tr><tr><td>00</td><td>April 24, 2020</td><td>Chris Goulopoulos</td><td>Michael Godfrey Other CTSC/D/M participants</td><td>Updated to reflect comments from review</td></tr><tr><td>01</td><td>July 9, 2020</td><td>Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td></tr><tr><td>02</td><td>August 27, 2020</td><td>Mark Salsberg</td><td>Chris Goulopoulos</td><td>Minor corrections and edits</td></tr><tr><td>03</td><td>June 29, 2021</td><td>. Chris Goulopoulos_</td><td>| Mark Salsberg</td><td>Adjusted wording for responsibilities based on CTS contractual agreements</td></tr></table>')],
 'output_text': 'The

In [ ]:
from unstructured.chunking.title import chunk_by_title

In [158]:
images = [el for el in elements if el.to_dict()['type'] == 'Image']

for i in images:
    print(i)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
     
  
Baseline Configuration Struc
Cyber Asset Baseline Configuration List |Review/Update Date: aor Document Classification Level: Cybersecurity Information - Confidential ‘Configuration Change st = pom i a — A cert eta pan EM Pin et ug “ 23 _|LAN Switch [Network switch '- Network Switches 24_|LAN Switch 

In [152]:
doc_text = '' 
for el in elements:
    if el.to_dict()['type'] != 'Header' and el.to_dict()['type'] != 'Footer':
        if el.to_dict()['type'] == 'Table':
            table_html = el.metadata.text_as_html

            parser = etree.XMLParser(remove_blank_text=True)
            file_obj = StringIO(table_html)
            tree = etree.parse(file_obj, parser)
            print(etree.tostring(tree, pretty_print=True).decode())
            doc_text += etree.tostring(tree, pretty_print=True).decode() + '\n'
        else:
            doc_text += el.to_dict()['text'] + '\n'

<table>
  <thead>
    <th>Revision</th>
    <th>Revision Date</th>
    <th>Author(s)</th>
    <th>Reviewer(s)</th>
    <th>Description of Changes</th>
  </thead>
  <tr>
    <td>PA</td>
    <td>April 8, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Initial draft for review</td>
  </tr>
  <tr>
    <td>00</td>
    <td>April 24, 2020</td>
    <td>Chris Goulopoulos</td>
    <td>Michael Godfrey Other CTSC/D/M participants</td>
    <td>Updated to reflect comments from review</td>
  </tr>
  <tr>
    <td>01</td>
    <td>July 9, 2020</td>
    <td>Chris Goulopoulos_</td>
    <td>| Mark Salsberg</td>
    <td>Updated to reflect comments from 2700- 67ASA1-01-CTSC-0012_02.pdf</td>
  </tr>
  <tr>
    <td>02</td>
    <td>August 27, 2020</td>
    <td>Mark Salsberg</td>
    <td>Chris Goulopoulos</td>
    <td>Minor corrections and edits</td>
  </tr>
  <tr>
    <td>03</td>
    <td>June 29, 2021</td>
    <td>. Chris Goulopoulos_</td>
    <td>| Mark

In [153]:
print(doc_text)

CROSSLIN TRANSIT SOLUTIONS
Oct 20, 2021 
Cyber Security – Change Control and Configuration Management Process
Prepared by:   Reviewed by:   Reviewed by:   Approved by:   Accepted by:   For   Eglinton Crosstown LRT   Chris Goulopoulos  Consultant, CTS   William VanRyswyk, CPP  Chief Security Officer, CTS   Evan Harben   Cyber Security & OT Network  Specialist, CTSC  Hadi Meshgi   Technical Lead - Communication &  Cybersecurity, CTSC   Oct 20, 2021  Kyle Booth Systems Integration  Authority, CTSC   Name, Title   Signature/Date   Document No.   5000-00-WGD-48PA-1020   Rev.   03   July 26, 2021  
if iA vo hay ~ Gi | x 4 July Lp >
Nee gle “dr
© 2021 Crosslinx Transit Solutions This document is licensed to Crosslinx Transit Solutions and cannot be used, reproduced, published, and/or revealed without prior written authorization. This document has been classified as Cyber Security - Confidential. Unauthorized use or distribution is prohibited. Printed copies of this document are uncontrolled. 

In [154]:
chatInstance = ChatOpenAI(temperature=0, openai_api_key=openai_api_key, model_name="gpt-4o")
chatInstance.get_num_tokens(doc_text)

14452

In [155]:
response = chatInstance.invoke(f"Clean the following document to be embedded in a vector database. Return only the cleaned document text: {doc_text}")

In [157]:
print(response.content)

Crosslinx Transit Solutions (CTS) has implemented a Cyber Security Plan (CSP) and associated policies and processes for protecting the Eglinton Crosstown Light Rail Transit (ECLRT’s) Cyber Systems from cyber security threats that could compromise the safe and reliable operation of the ECLRT.

The CTS Cyber Security - Management Controls Policy, establishes the management controls and the framework for cyber security standards, plans and processes for securing those Cyber Systems that are essential for the safe and reliable operation of the ECLRT.

The Cyber Security – Change Control and Configuration Management Standard, defined the cyber security controls and required processes for managing changes to ECLRT’s operational Cyber Systems and associated Cyber Assets and their configurations.

This document presents the processes for implementing, sustaining, and adhering to the cyber security controls for managing and controlling changes to ECLRT’s Cyber Systems and changes to the baselin

#### Conclusions
**Good Information**  
It seems like we can extract all the information with a good deal of clarity. Combining Unstructured with GPT-4o could prove incredibly useful. Basically, we would extract all the chunks, and then, for all chunks that arent headers or footers, we would pass the elements text property (text_as_html property for tables) to GPT-4o to clean the text for text elements, and summarize the tables for table elements. We would then concatenate all the information into one document, then pass that to GPT-4o to pull Title, Doc type, section headings, topics, authors, revision dates, 